lib/semian/adapter.rb

```diff
@@ -0,0 +1,34 @@
+module Semian
+  module Adapter
+    def semian_identifier
+      raise NotImplementedError.new("Semian adapters must implement a `semian_identifier` method")
+    end
+
+    private
+
+    def acquire_semian_resource(scope:)
+      return yield if @resource_acquired # Prevent a single adapter from acquiring the resource twice by nesting calls
```

Comment:
> Does pipeline actually end up acquiring a connection as well?

Pipelined is not a subclass of `Redis::Client`. Also the pipepline are first evaluated and then the client send the commands to the server, e.g:

```ruby
def pipelined
  yield commands_collector
  ensure_connected do
    perform commands_collector.commands
  end
end
```
(This is pseudo code but the idea is here)

> Or even if you could every do something like:

That would be fine too since the flag is an instance variable of the adapter, in this case each redis has his own flag.